1_fetch/src/fetch_nhdv2_attributes_from_sb.R

```diff
@@ -0,0 +1,163 @@
+  }
+  
+  # 2) Download data from ScienceBase
+  out_file <- download_sb_file(sb_id = unique(vars_item$sb_id),
+                               file_name = c(item_names),
+                               out_dir = save_dir)
+  
+  # 3) Select column names to be retained from original downloaded data
+  col_names <- c("COMID",
+                 vars_item$Catchment.Item.Name,
+                 vars_item$Divergence.Item.Name,
+                 vars_item$Total.Upstream.Item.Name)
+  col_names <- col_names[!is.na(col_names)]
+  
+  # Reformat column names for certain datasets
+  # National Inventory of Dams data:
+  if(unique(vars_item$sb_id) == "58c301f2e4b0f37a93ed915a"){
+    years <- str_extract(out_file,"\\d{2,}")
+    col_names <- format_col_names_years(col_names,years,yr_pattern = "YYYY")
+  }
+  
+  # HDENS data:
+  if(unique(vars_item$sb_id) == "5910de31e4b0e541a03ac983"){
+    years <- str_extract(out_file,"\\d{2,}")
+    col_names <- format_col_names_years(col_names,years,yr_pattern = "XX")
+  }
+  
+  # monthly average precipitation data (CAT,ACC,and TOT have different sb_id's):
+  if(unique(vars_item$sb_id) %in% c("5734acafe4b0dae0d5de622d","5730f062e4b0dae0d5db1fbe","573362bce4b0dae0d5dd6193")){
+    years <- str_extract(out_file,"\\d{2,}")
+    col_names <- format_col_names_years(col_names,years,yr_pattern = "YYYY")
+  }
+  
+  # 4) Unzip out_files, filter to COMIDs of interest, and return combined data frame
+  data_out <- lapply(out_file,unzip_and_clip_sb_data,
+                     col_names = col_names,
+                     comids = comids,
+                     save_dir = save_dir,
+                     delete_local_copies = delete_local_copies) %>%
+    Reduce(full_join,.) %>% 
+    suppressMessages() %>%
+    suppressWarnings()
+  
+  # 5) Save file
+  data_out_path <- paste0(save_dir,"/",unique(vars_item$Dataset_name),".csv")
+  write_csv(data_out,file = data_out_path)
+  
+  return(data_out_path)
+  
+}
+
+
+unzip_and_clip_sb_data <- function(out_file,col_names,comids,save_dir,delete_local_copies = TRUE){
+  #'
+  #' @description This function unzips zipped files from ScienceBase, reads in the unzipped data table, 
+  #' and filters the CONUS-scale data to retain the NHDPlusV2 COMID's of interest.
+  #' 
+  #' @param out_file character string indicating the file path of the zipped data
+  #' @param col_names string vector containing the columns to return
+  #' @param comids vector of COMIDs to retain from CONUS-scale datasets
+  #' @param save_dir character string indicting the file path to save the unzipped data
+  #' @param delete_local_copies logical; indicates whether to delete CONUS-scale zipped/unzipped data copies from save_dir
+  #' delete_local_copies defaults to TRUE
+  #' 
+  
+  # Unzip downloaded files
+  unzip(zipfile = out_file, exdir = save_dir, overwrite = TRUE)
+  
+  # Parse name(s) of unzipped files
+  file_name <- basename(out_file)
+  file_name_sans_ext <- str_split(file_name,".[[:alnum:]]+$")[[1]][1]
```

Comment:
[randomly skimming github stuff and this caught my eye] - do you have a reason not to use the built-in `tools::file_path_sans_ext()` here? This `str_split` approach works for 1 file but not multiple files, so needs to be modified in some way if you're serious about the plural option implied by `name(s)` 3 lines above

Reply:
Thanks for that suggestion, Alison! In short, I don't have a good reason not to use the built-in function you reference. My brain got going on regex I guess 😄. We map over this function for individual files so I don't expect that the `str_split` approach would cause issues, but it's good to simplify the code (and the plural option that was implied in the commented line above). I've edited this line to use `tools::file_path_sans_ext()` instead.